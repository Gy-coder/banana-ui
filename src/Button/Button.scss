$h: 32px;
$border-color: #d9d9d9;
$color: #333;
$blue: #40a9ff;
$radius: 4px;
$red: red;
$grey: grey;
.g-button {
  box-sizing: border-box;
  padding: 0 12px;
  height: $h;
  cursor: pointer;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  white-space: nowrap;
  background: white;
  color: $color;
  border: 1px solid $border-color;
  border-radius: $radius;
  box-shadow: 0 1px 0 fade-out(black, 0.95);
  & + & {
    margin-left: 8px;
  }
  &:hover,
  &:focus {
    color: $blue;
    border-color: $blue;
  }
  &:focus {
    outline: none;
  }
  &::-moz-focus-inner {
    border: 0;
  }
  &.g-theme-link {
    border-color: transparent;
    box-shadow: none;
    color: $blue;
    &:hover,
    &:focus {
      color: lighten($blue, 10%);
    }
  }
  &.g-theme-text {
    border-color: transparent;
    box-shadow: none;
    color: inherit;
    &:hover,
    &:focus {
      background: darken(white, 5%);
    }
  }
  &.g-size-big {
    font-size: 24px;
    height: 48px;
    padding: 0 16px;
  }
  &.g-size-small {
    font-size: 12px;
    height: 20px;
    padding: 0 4px;
  }
  &.g-theme-button {
    &.g-level-main {
      background: $blue;
      color: white;
      border-color: $blue;
      &:hover,
      &:focus {
        background: darken($blue, 10%);
        border-color: darken($blue, 10%);
      }
    }
    &.g-level-danger {
      background: $red;
      border-color: $red;
      color: white;
      &:hover,
      &:focus {
        background: darken($red, 10%);
        border-color: darken($red, 10%);
      }
    }
  }
  &.g-theme-link {
    &.g-level-danger {
      color: $red;
      &:hover,
      &:focus {
        color: darken($red, 10%);
      }
    }
  }
  &.g-theme-text {
    &.g-level-main {
      color: $blue;
      &:hover,
      &:focus {
        color: darken($blue, 10%);
      }
    }
    &.g-level-danger {
      color: $red;
      &:hover,
      &:focus {
        color: darken($red, 10%);
      }
    }
  }
  &.g-theme-button {
    &[disabled] {
      cursor: not-allowed;
      color: $grey;
      &:hover {
        border-color: $grey;
      }
    }
  }
  &.g-theme-link,
  &.g-theme-text {
    &[disabled] {
      cursor: not-allowed;
      color: $grey;
    }
  }
  > .g-loadingIndicator {
    box-sizing: border-box;
    width: 14px;
    height: 14px;
    display: inline-block;
    margin-right: 4px;
    border-radius: 8px;
    border-color: $blue $blue $blue transparent;
    border-style: solid;
    border-width: 2px;
    animation: g-spin 1s infinite linear;
  }
}
@keyframes g-spin {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
