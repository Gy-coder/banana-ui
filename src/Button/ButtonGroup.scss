$h: 32px;
$border-color: #d9d9d9;
$color: #333;
$blue: #40a9ff;
$radius: 4px;
$red: red;
$grey: grey;
.g-button-group {
  display: inline-flex;
  vertical-align: center;
  > .g-button {
    border-radius: 0;
    margin-left: -1px !important;
    &:hover {
      z-index: 1;
    }
    &:focus {
      z-index: 1;
    }
    &:first-child {
      border-top-left-radius: $radius;
      border-bottom-left-radius: $radius;
    }
    &:last-child {
      border-top-right-radius: $radius;
      border-bottom-right-radius: $radius;
    }
  }
  > .g-button + .g-button {
    margin: 0;
  }
}
