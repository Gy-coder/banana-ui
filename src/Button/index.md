---
title: Button
nav:
  title: 基础组件
  path: /components
---

# Button

按钮一般用于点击触发一个操作

# 何时使用

标记了一个（或封装一组）操作命令，响应用户点击行为，触发相应的业务逻辑。

在 banana-ui 中我们提供了三种按钮。

- 默认按钮：用于没有主次之分的一组行动点。

- 文本按钮：用于最次级的行动点。

- 链接按钮：用于作为外链的行动点。

提供了大中小三个不同的尺寸的按钮

以及四种状态属性与上面配合使用。

- 危险：删除/移动/修改权限等危险操作，一般需要二次确认。

- 主按钮: 用于主行动点，一个操作区域只能有一个主按钮。

- 禁用：行动点不可用的时候，一般需要文案解释。

- 加载中：用于异步操作等待反馈的时候，也可以避免多次提交。

# 代码演示

<code src="./demos/basic.tsx" />
<code src="./demos/size.tsx" />
<code src="./demos/level.tsx" />
<code src="./demos/disabled.tsx">
<code src="./demos/loading.tsx" />

<API src="./Button.tsx"/>
