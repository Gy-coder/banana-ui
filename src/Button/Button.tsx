import React from 'react';
import classnames from 'classnames';
import './Button.scss';

interface ButtonProps
  extends Omit<React.ButtonHTMLAttributes<HTMLButtonElement>, 'disabled'> {
  /**
   * 可以这样写属性描述
   * @description     Button的类型
   * @default         button
   */
  theme?: 'button' | 'link' | 'text';
  /**
   * 可以这样写属性描述
   * @description      Button的大小
   * @default          normal
   */
  size?: 'small' | 'normal' | 'big';
  /**
   * 可以这样写属性描述
   * @description      Button的级别
   * @default          normal
   */
  level?: 'main' | 'danger' | 'normal';
  /**
   * 可以这样写属性描述
   * @description      是否禁用
   * @default          false
   */
  disabled?: boolean;
  /**
   * 可以这样写属性描述
   * @description      是否显示加载状态
   * @default          false
   */
  loading?: Boolean;
}

const Button: React.FC<ButtonProps> = (props) => {
  const {
    children,
    className,
    theme,
    size,
    level,
    disabled,
    loading,
    ...rest
  } = props;
  const classes = classnames('g-button', className, {
    [`g-theme-${theme}`]: theme,
    [`g-size-${size}`]: size,
    [`g-level-${level}`]: level,
  });
  return (
    <button className={classes} disabled={disabled} {...rest}>
      {loading && <span className="g-loadingIndicator" />}
      {children}
    </button>
  );
};

Button.defaultProps = {
  theme: 'button',
  size: 'normal',
  level: 'normal',
  disabled: false,
  loading: false,
};

export default Button;
