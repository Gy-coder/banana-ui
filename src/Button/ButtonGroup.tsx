import React from 'react';
import Button, { ButtonProps } from './Button';
import './ButtonGroup.scss';

const ButtonGroup: React.FC = (props) => {
  const { children } = props;
  React.Children.map(children, (child) => {
    const childElement = child as React.FunctionComponentElement<ButtonProps>;
    if (childElement.type !== Button) {
      throw new Error('CheckboxGroup的子元素必须是Checkbox');
    }
  });
  return <div className="g-button-group">{children}</div>;
};

export default ButtonGroup;
