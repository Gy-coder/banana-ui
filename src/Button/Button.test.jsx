import React from 'react';
import {
  render,
  act,
  cleanup,
  screen,
  fireEvent,
} from '@testing-library/react';
import Button from './Button';

afterEach(cleanup);

it('renders correctly', () => {
  const { asFragment } = render(<Button>Click Me</Button>);
  expect(screen.getByText(/Click Me/i)).toBeTruthy();
  expect(asFragment()).toMatchSnapshot();
});

it('accept onClick and rest props', () => {
  const handleClick = jest.fn();
  render(
    <Button data-testid="button" onClick={handleClick}>
      Click
    </Button>,
  );
  act(() => {
    fireEvent.click(screen.getByTestId('button'));
  });
  expect(handleClick).toHaveBeenCalledTimes(1);
});

it('should be disabled', () => {
  const handleClick = jest.fn();
  render(
    <Button theme="link" disabled onClick={handleClick}>
      Click
    </Button>,
  );
  act(() => {
    fireEvent.click(screen.getByRole('button'));
  });
  expect(handleClick).not.toHaveBeenCalled();
});
