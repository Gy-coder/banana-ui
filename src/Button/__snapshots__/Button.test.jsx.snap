// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<DocumentFragment>
  <button
    class="g-button g-theme-button g-size-normal g-level-normal"
  >
    Click Me
  </button>
</DocumentFragment>
`;
